### PR TITLE
[Numpy] add: numpy op trilindices

### PR DESCRIPTION
**Pull request by BenjaminCHEN2016**
## Description ##
(Brief description on what this PR is about)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by BenjaminCHEN2016** [comment]
@mxnet-bot run ci [all]

**Comment by haojin2** [comment]
@BenjaminCHEN2016 please remember to do `git submodule update --init --recursive` after you rebase to get rid of the change to `3rdparty/mkldnn`

File: tests/python/unittest/test_numpy_op.py [hunk 1]
```diff
@@ -6693,6 +6693,35 @@ def test_np_builtin_op_signature():
         assert op is not None
         assert str(op.__signature__) == str(inspect.signature(_op_from_doc))
 
+@with_seed()
```
Comment (haojin2) [inline tests/python/unittest/test_numpy_op.py]: an extra blank line above

File: tests/python/unittest/test_numpy_op.py [hunk 2]
```diff
@@ -6693,6 +6693,35 @@ def test_np_builtin_op_signature():
...
+                for hybridize in [True, False]:
+                    # dummy nparray for hybridize
+                    x = mx.nd.ones((1,1)).as_np_ndarray()
+                    test_trilindices = TestTrilindices(n, k, m)
+                    if hybridize:
+                        test_trilindices.hybridize()
+                    mx_out = test_trilindices(x)[1]
+                    assert len(mx_out) == 2
+                    assert same(mx_out[0], np_out[0])
+                    assert same(mx_out[1], np_out[1])
```
Comment (haojin2) [inline tests/python/unittest/test_numpy_op.py]: an extra blank line below.

File: tests/python/unittest/test_numpy_op.py [hunk 3]
```diff
@@ -6693,6 +6693,35 @@ def test_np_builtin_op_signature():
...
+        def hybrid_forward(self, F, x, *args, **kwargs):
+            return x, F.np.tril_indices(n=self._n, k=self._k, m=self._m)
+    
+    for n in _np.random.random_integers(-10, 50, 10):
+        for k in _np.random.random_integers(-50, 50, 4):
+            for m in _np.random.random_integers(-10, 50, 4):
+                np_out = _np.tril_indices(n, k, m)
+                for hybridize in [True, False]:
+                    # dummy nparray for hybridize
+                    x = mx.nd.ones((1,1)).as_np_ndarray()
```
Comment (haojin2) [inline tests/python/unittest/test_numpy_op.py]: simply do `x = np.ones((1, 1))`

File: tests/python/unittest/test_numpy_op.py [hunk 4]
```diff
@@ -6693,6 +6693,35 @@ def test_np_builtin_op_signature():
...
+            self._n = n;
+            self._k = k;
+            if m is None:
+                m = n
+            self._m = m
+        
+        def hybrid_forward(self, F, x, *args, **kwargs):
+            return x, F.np.tril_indices(n=self._n, k=self._k, m=self._m)
+    
+    for n in _np.random.random_integers(-10, 50, 10):
```
Comment (haojin2) [inline tests/python/unittest/test_numpy_op.py]: how long is this unit test currently taking? If it's taking more than 3 seconds for a single run, please consider reducing those test cases.

File: src/operator/numpy/np_tril_indices_op.cc [hunk 5]
```diff
@@ -0,0 +1,46 @@
+/*
```
Comment (haojin2) [inline src/operator/numpy/np_tril_indices_op.cc]: move this op to `np_matrix_op` files

File: python/mxnet/symbol/numpy/_symbol.py [hunk 6]
```diff
@@ -2197,6 +2197,87 @@ def tril(m, k=0):
     """
     return _npi.tril(m, k)
 
+@set_module('mxnet.symbol.numpy')
```
Comment (haojin2) [inline python/mxnet/symbol/numpy/_symbol.py]: one more blank line above.

File: python/mxnet/symbol/numpy/_symbol.py [hunk 7]
```diff
@@ -2197,6 +2197,87 @@ def tril(m, k=0):
...
+    >>> a
+    array([[-10, -10, -10,   3],
+           [-10, -10, -10, -10],
+           [-10, -10, -10, -10],
+           [-10, -10, -10, -10]])
+
+    """
+    if m is None:
+        m = n
+    return _npi.tril_indices(n, k, m)
```
Comment (haojin2) [inline python/mxnet/symbol/numpy/_symbol.py]: one more blank line below

File: python/mxnet/numpy/multiarray.py [hunk 8]
```diff
@@ -5533,6 +5533,87 @@ def tril(m, k=0):
     """
     return _mx_nd_np.tril(m, k)
 
+@set_module('mxnet.numpy')
```
Comment (haojin2) [inline python/mxnet/numpy/multiarray.py]: one more blank line above.

File: python/mxnet/numpy/multiarray.py [hunk 9]
```diff
@@ -5533,6 +5533,87 @@ def tril(m, k=0):
...
+    >>> a
+    array([[-10, -10, -10,   3],
+           [-10, -10, -10, -10],
+           [-10, -10, -10, -10],
+           [-10, -10, -10, -10]])
+
+    """
+    if m is None:
+        m = n
+    return _mx_nd_np.tril_indices(n, k, m)
```
Comment (haojin2) [inline python/mxnet/numpy/multiarray.py]: one more blank line below

File: python/mxnet/numpy/multiarray.py [hunk 10]
```diff
@@ -5533,6 +5533,87 @@ def tril(m, k=0):
...
+        indices will be valid.
+    k : int, optional
+        Diagonal offset (see `tril` for details).
+    m : int, optional
+        .. versionadded:: 1.9.0
+
+        The column dimension of the arrays for which the returned
+        arrays will be valid.
+        By default `m` is taken equal to `n`.
+
```
Comment (haojin2) [inline python/mxnet/numpy/multiarray.py]: get rid of this line.

File: python/mxnet/numpy/multiarray.py [hunk 11]
```diff
@@ -5533,6 +5533,87 @@ def tril(m, k=0):
...
+    See also
+    --------
+    triu_indices : similar function, for upper-triangular.
+    mask_indices : generic function accepting an arbitrary mask function.
+    tril, triu
+
+    Notes
+    -----
+    .. versionadded:: 1.4.0
```
Comment (haojin2) [inline python/mxnet/numpy/multiarray.py]: no need for `Notes` section

File: python/mxnet/numpy/multiarray.py [hunk 12]
```diff
@@ -5533,6 +5533,87 @@ def tril(m, k=0):
...
+           [-1, -1,  6,  7],
+           [-1, -1, -1, 11],
+           [-1, -1, -1, -1]])
+
+    These cover almost the whole array (two diagonals right of the main one):
+
+    >>> a[il2] = -10
+    >>> a
+    array([[-10, -10, -10,   3],
+           [-10, -10, -10, -10],
```
Comment (haojin2) [inline python/mxnet/numpy/multiarray.py]: remember to verify all of those examples work for your implementation.

File: python/mxnet/symbol/numpy/_symbol.py [hunk 13]
```diff
@@ -2197,6 +2197,87 @@ def tril(m, k=0):
...
+        .. versionadded:: 1.9.0
+
+        The column dimension of the arrays for which the returned
+        arrays will be valid.
+        By default `m` is taken equal to `n`.
+
+
+    Returns
+    -------
+    inds : tuple of arrays
```
Comment (haojin2) [inline python/mxnet/symbol/numpy/_symbol.py]: here it should be `tuple of _Symbol` I guess

File: python/mxnet/numpy/multiarray.py (hunk 12 above)
Comment (BenjaminCHEN2016) [reply python/mxnet/numpy/multiarray.py]: I have added more tests.

File: src/operator/numpy/np_matrix_op.cc [hunk 14]
```diff
@@ -1115,6 +1115,66 @@ NNVM_REGISTER_OP(_backward_np_dstack)
...
+  CHECK_EQ(out_attrs->size(), 2U);
+
+  TYPE_ASSIGN_CHECK(*out_attrs, 0, mshadow::kInt64);
+  TYPE_ASSIGN_CHECK(*out_attrs, 1, mshadow::kInt64);
+
+  return true;
+}
+
+inline bool TrilindicesOpShape(const nnvm::NodeAttrs& attrs,
+                            mxnet::ShapeVector* in_attrs,
```
Comment (haojin2) [inline src/operator/numpy/np_matrix_op.cc:1133]: alignment

File: src/operator/numpy/np_matrix_op.cc [hunk 15]
```diff
@@ -1115,6 +1115,66 @@ NNVM_REGISTER_OP(_backward_np_dstack)
...
+                            mxnet::ShapeVector* out_attrs) {
+  CHECK_EQ(in_attrs->size(), 0U);
+  CHECK_EQ(out_attrs->size(), 2U);
+
+  const NumpyTrilindicesParam& param =
+                              nnvm::get<NumpyTrilindicesParam>(attrs.parsed);
```
Comment (haojin2) [inline src/operator/numpy/np_matrix_op.cc:1139]:
```c++
  const NumpyTrilindicesParam& param =
    nnvm::get<NumpyTrilindicesParam>(attrs.parsed);
```

File: src/operator/numpy/np_matrix_op-inl.h [hunk 16]
```diff
@@ -287,6 +287,106 @@ void NumpyVstackBackward(const nnvm::NodeAttrs& attrs,
...
+  MSHADOW_XINLINE static void Map(int i, DType* out_data0, DType* out_data1,
+                                  int* data, int length) {
+    KERNEL_ASSIGN(out_data0[i], req, data[i]);
+    KERNEL_ASSIGN(out_data1[i], req, data[i + length]);
+  }
+};
+
+template<typename xpu>
+void TrilindicesOpForward(const nnvm::NodeAttrs& attrs,
+                    const OpContext& ctx,
```
Comment (haojin2) [inline src/operator/numpy/np_matrix_op-inl.h:329]: alignment

**Comment by BenjaminCHEN2016** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by BenjaminCHEN2016** [comment]
@mxnet-bot run ci [macosx]

**Comment by BenjaminCHEN2016** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by BenjaminCHEN2016** [comment]
@mxnet-bot run ci [unix-gpu]

**Pull request merged**